lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -594,6 +727,40 @@ static LogicalResult verifyWiresOp(WiresOp wires) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// CombGroupOp
+//===----------------------------------------------------------------------===//
+
+/// Verifies a combinational group may contain only combinational primitives.
+// TODO(www.github.com/llvm/circt/issues/1739): Add Combinational trait.
```

Comment:
I don’t know if this is correct. For example, it is perfectly okay to read a value from a register in a combinational group. Similarly, memories may be used in combinational groups as well if they’re being read.

Reply:
Can you provide some more insight? I'm happy to remove this, but there is some misunderstanding on my part. In Calyx, I know we always consider register writes to be one cycle for simplicity, given that it is a non-combinational circuit. For memory, we consider memory writes to be a single cycle as well. I imagine a memory load would *at least* take a single cycle. So to read from a memory and then use it in a combinational group, the control might look like:

```mlir
calyx.control {
  calyx.seq {
    calyx.enable @LoadFromMemory {static=N} // Read the intended value.
    calyx.if @CombGroup { ... }             // Use the value combinationally.
  }
}
```

In other words, why can we get away with using non-combinational modules in combinational groups?

Reply:
The problem with this verification check is that while combinational primitives (`comb primitive`) only have combinational circuit paths in them, a non-combinational primitive can have *both* combinational and sequential paths in it.

For example, the implementation of the memory primitive in Calyx takes one cycle to write a value into memory but can combinationally read the memory. So the following combinational groups are all fine (and indeed show up when generating code from Dahlia):

```
comb group read_reg {
  eq.left = reg.out;
  eq.right = 32'd10;
}

comb group read_mem {
  mem.addr0 = 4'd2;
  eq.left = mem.read_data;
  eq.right = 32'd10;
}
```

It is worth noting that different memory primitives may implement reads sequentially (i.e. take one or more cycles to return the value of a read). 

TLDR: Non-combinational primitives may be used in combinational groups as long as the ports they're been read from are on a combinational circuit path.

Reply:
Ok so just to clarify, you're implying that current memory primitives in Calyx (RegisterOp, MemoryOp) conduct reads combinationally.

Reply:
Registers always provide reads combinationally. Memories may provide reads combinationally.

Reply:
Fixed this. Thanks for the context.